File: src/main/java/com/github/nsc/de/shake/parser/node/variables/VariableAddAssignmentNode.java
package com.github.nsc.de.shake.parser.node.variables;

import com.github.nsc.de.shake.lexer.characterinput.position.PositionMap;
import com.github.nsc.de.shake.parser.node.Node;
import com.github.nsc.de.shake.parser.node.ValuedNode;

public class VariableAddAssignmentNode extends ValuedNode {

    private final ValuedNode variable;
    private final Node value;
    private final int operatorPosition;

    public VariableAddAssignmentNode(PositionMap map, ValuedNode variable, Node value, int operatorPosition) {
        super(map);
        this.variable = variable;
        this.value = value;
        this.operatorPosition = operatorPosition;
    }

    public ValuedNode getVariable() {
        return variable;
    }
    public Node getValue() {
        return value;
    }
    public int getOperatorPosition() {
        return operatorPosition;
    }

    @Override
    public String toString() {
        return '{' + getValue().toString() + "+=" + this.getValue().toString() + '}';
    }
}

File: parser/src/main/kotlin/com/github/nsc/de/shake/parser/node/variables/VariableAddAssignmentNode.kt
package com.github.nsc.de.shake.parser.node.variables

import com.github.nsc.de.shake.parser.node.*
import com.github.nsc.de.shake.util.characterinput.position.PositionMap

class VariableAddAssignmentNode(
    map: PositionMap,
    val variable: ValuedNode,
    val value: Node,
    val operatorPosition: Int
) : ValuedNode(map) {
    override fun toString(): String = "{$value+=$value}"
}